feat: Add Board Game tracking with BoardGameGeek integration

zskemp opened the pull request:
## 📋 Summary
Adds support for tracking board games using the BoardGameGeek (BGG) XML API v2, including search, metadata retrieval, and play tracking functionality.

## ✨ Features Added
- **New Media Type**: Board games (`boardgame`) with "Plays" tracking unit
- **BGG API Integration**: Full XML API v2 support with Bearer token authentication
- **Search with Pagination**: 20 results per page with image thumbnails
- **Game Metadata**: Title, image, description, year, player count, playtime, age rating, BGG rating
- **Play Tracking**: Track number of times played (no max_progress limit)
- **Two-tier Caching**: 
  - Game IDs cached for 24 hours (reduces API calls)
  - Page results cached for 24 hours (instant pagination)

## 🔧 Technical Changes

**New Files:**
- `src/app/providers/bgg.py` - BoardGameGeek API provider implementation

**Modified Files:**
- `src/app/models.py` - Added `BOARDGAME` media type and `BoardGame` model
- `src/app/forms.py` - Added `BoardgameForm` for tracking UI
- `src/app/media_type_config.py` - Board game configuration (icon, colors, units)
- `src/app/providers/services.py` - Integrated BGG provider into search/metadata routing
- `src/config/settings.py` - Added `BGG_API_TOKEN` environment variable
- `README.md` - Updated feature list and added BGG token to `.env` example
- `requirements.txt` - No new dependencies (uses built-in `xml.etree.ElementTree`)

**Database Migration:**
- Included in existing migration file for `BoardGame` model

## 🎮 How It Works

**Search Flow:**
1. User searches for "catan" → BGG search API returns 141 game IDs
2. System fetches images for first 20 games (page 1)
3. Results cached for 24 hours
4. Pagination loads additional pages (cached after first fetch)

**Tracking:**
1. Click "Track" → BGG metadata API fetches full game details
2. User can set status (Planning/Playing/Completed) and track plays
3. Progress = number of plays (unlimited)

**Rate Limiting:** 
- No artificial delays (BGG handles server-side rate limiting with token auth)
- Respects BGG's authorization requirements

## 🔑 Configuration Required

Users need to set the `BGG_API_TOKEN` environment variable:

1. Visit https://boardgamegeek.com/using_the_xml_api
2. Request API access (requires BGG account)
3. Wait for approval (~1-2 business days)
4. Add token to `.env`: `BGG_API_TOKEN=your_token_here`

## 📝 Wiki Update Needed

The [Environment Variables wiki page](https://github.com/FuzzyGrim/Yamtrack/wiki/Environment-Variables) should be updated with:

```markdown
### BGG_API_TOKEN
**Required for:** Board game tracking via BoardGameGeek

**How to get it:**
1. Visit https://boardgamegeek.com/using_the_xml_api
2. Log in with your BGG account
3. Request API access and wait for approval (usually takes 1-2 business days)
4. Once approved, you'll receive your Bearer token
5. Add it to your `.env` file as `BGG_API_TOKEN=your_token_here`

**Note:** As of November 2025, BGG requires authorization for all XML API v2 requests.
```

## ✅ Testing Performed
- [x] Search functionality with pagination (141 results for "catan")
- [x] Image loading for all game types (base games, expansions, accessories)
- [x] Metadata retrieval with all fields populated
- [x] Track/save board games with play counts
- [x] Form validation and submission
- [x] Cache performance (instant page loads on cache hit)
- [x] BGG API authorization handling

## 🎯 Related Issues
Closes #967 : Addresses the need for board game tracking in Yamtrack's media library.

## 📸 Screenshots
<img width="3778" height="1967" alt="Screenshot 2025-11-18 105826" src="https://github.com/user-attachments/assets/44d05406-5829-4c32-9357-2b20fa198b5c" />
<img width="3805" height="1956" alt="Screenshot 2025-11-18 105926" src="https://github.com/user-attachments/assets/a8d0ee8f-26e8-45d7-91d6-7e29879531b9" />
<img width="3763" height="1936" alt="Screenshot 2025-11-18 105956" src="https://github.com/user-attachments/assets/8415dd31-ecef-4a5a-8103-64e15f565f92" />

zacharyskemp committed:
Board Game Category Feature

zacharyskemp committed:
fix: Replace xml.etree with defusedxml for secure XML parsing (B314)

dannyvfilms commented:
I'm open to integrating this into my fork, but it might take me a bit to review.

66Bunz commented on src/app/providers/bgg.py:
```diff
@@ -0,0 +1,301 @@
+        "User-Agent": "Yamtrack/1.0 (https://github.com/FuzzyGrim/Yamtrack)"
+    }
+    
+    # BGG now requires Bearer token authorization (as of Nov 2024)
```
I think this redundant comment can be removed

66Bunz commented on src/app/providers/bgg.py:
```diff
@@ -0,0 +1,301 @@
+    try:
+        response = requests.get(url, params=params, headers=headers, timeout=10)
+        
+        # Check for missing/invalid authorization
```
This comment can be removed, I think the code is self explaining

66Bunz commented on src/app/providers/bgg.py:
```diff
@@ -0,0 +1,301 @@
+        
+        response.raise_for_status()
+
+        # BGG may return 202 (queued request), need to retry
```
This comment can be removed, I think the code is self explaining

66Bunz commented on src/app/providers/bgg.py:
```diff
@@ -0,0 +1,301 @@
+            time.sleep(2)
+            return _bgg_request(endpoint, params)
+
+        # Parse XML response
```
This comment can be removed, I think the code is self explaining

66Bunz commented on src/app/providers/bgg.py:
```diff
@@ -0,0 +1,301 @@
+        }
+        root = _bgg_request("search", params)
+
+        # Collect game IDs and names from search results
```
This comment can be removed, I think the code is self explaining

66Bunz commented on src/app/providers/bgg.py:
```diff
@@ -0,0 +1,301 @@
+                game_names[game_id] = name_elem.get("value", "Unknown")
+        
+        game_data = {"ids": game_ids, "names": game_names}
+        # Cache game IDs for 1 day
```
This comment can be removed, I think the code is self explaining

66Bunz commented on src/app/providers/bgg.py:
```diff
@@ -0,0 +1,301 @@
+    game_ids = game_data["ids"]
+    game_names = game_data["names"]
+    
+    # Check if this specific page is cached
```
This comment can be removed, I think the code is self explaining

66Bunz commented on src/app/providers/bgg.py:
```diff
@@ -0,0 +1,301 @@
+        return cached_page
+    
+    # Paginate: 20 results per page
+    per_page = 20
```
This constant should be at the top of the file/function

zskemp replied:
Updated to this format. New constant can be found on top as RESULTS_PER_PAGE

66Bunz commented on src/app/providers/bgg.py:
```diff
@@ -0,0 +1,301 @@
+    if cached_page:
+        return cached_page
+    
+    # Paginate: 20 results per page
```
This comment can be removed

66Bunz commented on src/app/providers/bgg.py:
```diff
@@ -0,0 +1,301 @@
+        results=results,
+    )
+
+    # Cache this page for 1 day
```
This comment can be removed, I think the code is self explaining

66Bunz commented on src/app/providers/bgg.py:
```diff
@@ -0,0 +1,301 @@
+        "related": {},
+    }
+
+    # Cache for 7 days
```
This comment can be removed, I think the code is self explaining

66Bunz reviewed (changes requested):
I think this looks very good, but it needs to be updated with the latest changes.

I just added my personal take on some redundant comments.

makefu commented:
I would love to see this PR merged, it looks extremely promising!

dannyvfilms commented:
> I would love to see this PR merged, it looks extremely promising!

I have this running on my fork if you'd like to start using it.

makefu commented:
> > I would love to see this PR merged, it looks extremely promising!
> 
> I have this running on my fork if you'd like to start using it.

i am currently using published container at ghcr.io/fuzzygrim/yamtrack:latest , do you by chance also publish your containers?

dannyvfilms commented:
> i am currently using published container at ghcr.io/fuzzygrim/yamtrack:latest , do you by chance also publish your containers?

Yes, you can pull `ghcr.io/dannyvfilms/yamtrack:latest` to run my version. Be careful though, if you just switch your container a bunch of migrations will run that are a one-way street. I also have a demo you can try first if you want. It's linked on my repo.

zacharyskemp committed:
Board Game Feature with PR Feedback

zacharyskemp committed:
Board Game Feature with All PR Feedback

zskemp commented:
@66Bunz @FuzzyGrim, I have updated the PR to re-base with the latest changes. I have also addressed all of @66Bunz 's PR comments. Please let me know if you have any more feedback, I am happy to change anything y'all would like!

zacharyskemp committed:
fix: resolve ruff lint errors and config bug for boardgame search

zacharyskemp committed:
Update to migrations to reduce any merge conflicts with migrations

FuzzyGrim commented:
Thanks for the PR, it looks great!
